In [23]:
import pandas as pd 

# Visualization - Plotly
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import matplotlib.pyplot as plt 
import seaborn as sns 

#### **1.0 📊 Descriptive Analysis 🔍**

In [3]:
data = pd.read_csv(r"C:\Users\MSI\Fraud detction MLops Project\data\fraud_oracle.csv")

In [4]:
num_rows, num_columns = data.shape
print(f"The dataset contains {num_rows} rows and {num_columns} columns.")

The dataset contains 15420 rows and 33 columns.


In [5]:
data.head()

Month  WeekOfMonth  DayOfWeek    Make AccidentArea DayOfWeekClaimed  \
0   Dec            5  Wednesday   Honda        Urban          Tuesday   
1   Jan            3  Wednesday   Honda        Urban           Monday   
2   Oct            5     Friday   Honda        Urban         Thursday   
3   Jun            2   Saturday  Toyota        Rural           Friday   
4   Jan            5     Monday   Honda        Urban          Tuesday   

  MonthClaimed  WeekOfMonthClaimed     Sex MaritalStatus  ...  AgeOfVehicle  \
0          Jan                   1  Female        Single  ...       3 years   
1          Jan                   4    Male        Single  ...       6 years   
2          Nov                   2    Male       Married  ...       7 years   
3          Jul                   1    Male       Married  ...   more than 7   
4          Feb                   2  Female        Single  ...       5 years   

  AgeOfPolicyHolder PoliceReportFiled WitnessPresent AgentType  \
0          26 to 30                No             No  External   
1          31 to 35               Yes             No  External   
2          41 to 50                No             No  External   
3          51 to 65               Yes             No  External   
4          31 to 35                No             No  External   

   NumberOfSuppliments  AddressChange_Claim  NumberOfCars  Year  BasePolicy  
0                 none               1 year        3 to 4  1994   Liability  
1                 none            no change     1 vehicle  1994   Collision  
2                 none            no change     1 vehicle  1994   Collision  
3          more than 5            no change     1 vehicle  1994   Liability  
4                 none            no change     1 vehicle  1994   Collision  

[5 rows x 33 columns]

In [6]:
# Check for missing values
data.isnull().sum()

Month                   0
WeekOfMonth             0
DayOfWeek               0
Make                    0
AccidentArea            0
DayOfWeekClaimed        0
MonthClaimed            0
WeekOfMonthClaimed      0
Sex                     0
MaritalStatus           0
Age                     0
Fault                   0
PolicyType              0
VehicleCategory         0
VehiclePrice            0
FraudFound_P            0
PolicyNumber            0
RepNumber               0
Deductible              0
DriverRating            0
Days_Policy_Accident    0
Days_Policy_Claim       0
PastNumberOfClaims      0
AgeOfVehicle            0
AgeOfPolicyHolder       0
PoliceReportFiled       0
WitnessPresent          0
AgentType               0
NumberOfSuppliments     0
AddressChange_Claim     0
NumberOfCars            0
Year                    0
BasePolicy              0
dtype: int64

=> No missing data 

In [7]:
duplicate_rows = data.duplicated() # Check for duplicates
print(f"Number of duplicate rows: {duplicate_rows.sum()}")

Number of duplicate rows: 0


=> No duplicates

#### **1.1 Characteristic of Dataset**

In [8]:
# Function01: Summarize the characteristic of dataset

def summarize(DataFrame):
    
    summary = pd.DataFrame()
    
    # Data Type
    summary['Data Type'] = DataFrame.dtypes
    # N Unique
    summary['N Unique'] = DataFrame.nunique()
    # Unique
    summary['Unique'] = DataFrame.apply(lambda x: x.unique().tolist())
    # Max
    summary['Max'] = DataFrame.apply(lambda x: x.max() if pd.api.types.is_numeric_dtype(x) else '-')
    # Min
    summary['Min'] = DataFrame.apply(lambda x: x.min() if pd.api.types.is_numeric_dtype(x) else '-')
    
    # Measures of Central Tendency: Mean, Median, Mode 
    summary['Mean'] = DataFrame.apply(lambda x: round(x.mean(), 2) if pd.api.types.is_numeric_dtype(x) else '-')
    summary['Median'] = DataFrame.apply(lambda x: x.median() if pd.api.types.is_numeric_dtype(x) else '-')
    summary['Mode'] = DataFrame.apply(lambda x: x.mode().iloc[0] if not x.mode().empty else '-')
    
    # Measures of Dispersion: Range, Variance, Standard Deviation
    summary['Range'] = DataFrame.apply(lambda x: x.max() - x.min() if pd.api.types.is_numeric_dtype(x) else '-')
    summary['Variance'] = DataFrame.apply(lambda x: x.var() if pd.api.types.is_numeric_dtype(x) else '-')
    summary['Standard Deviation'] = DataFrame.apply(lambda x: x.std() if pd.api.types.is_numeric_dtype(x) else '-')
    
    # Measures of Shape: Skewness, Kurtosis
    summary['Skewness'] = DataFrame.apply(lambda x: round(x.skew(), 2) if pd.api.types.is_numeric_dtype(x) else '-')
    summary['Kurtosis'] = DataFrame.apply(lambda x: round(x.kurt(), 2) if pd.api.types.is_numeric_dtype(x) else '-')
    
    return summary

In [9]:
Summary = summarize(data)
Summary

Data Type  N Unique  \
Month                   object        12   
WeekOfMonth              int64         5   
DayOfWeek               object         7   
Make                    object        19   
AccidentArea            object         2   
DayOfWeekClaimed        object         8   
MonthClaimed            object        13   
WeekOfMonthClaimed       int64         5   
Sex                     object         2   
MaritalStatus           object         4   
Age                      int64        66   
Fault                   object         2   
PolicyType              object         9   
VehicleCategory         object         3   
VehiclePrice            object         6   
FraudFound_P             int64         2   
PolicyNumber             int64     15420   
RepNumber                int64        16   
Deductible               int64         4   
DriverRating             int64         4   
Days_Policy_Accident    object         5   
Days_Policy_Claim       object         4   
PastNumberOfClaims      object         4   
AgeOfVehicle            object         8   
AgeOfPolicyHolder       object         9   
PoliceReportFiled       object         2   
WitnessPresent          object         2   
AgentType               object         2   
NumberOfSuppliments     object         4   
AddressChange_Claim     object         5   
NumberOfCars            object         5   
Year                     int64         3   
BasePolicy              object         3   

                                                                 Unique  \
Month                 [Dec, Jan, Oct, Jun, Feb, Nov, Apr, Mar, Aug, ...   
WeekOfMonth                                             [5, 3, 2, 4, 1]   
DayOfWeek             [Wednesday, Friday, Saturday, Monday, Tuesday,...   
Make                  [Honda, Toyota, Ford, Mazda, Chevrolet, Pontia...   
AccidentArea                                             [Urban, Rural]   
DayOfWeekClaimed      [Tuesday, Monday, Thursday, Friday, Wednesday,...   
MonthClaimed          [Jan, Nov, Jul, Feb, Mar, Dec, Apr, Aug, May, ...   
WeekOfMonthClaimed                                      [1, 4, 2, 3, 5]   
Sex                                                      [Female, Male]   
MaritalStatus                        [Single, Married, Widow, Divorced]   
Age                   [21, 34, 47, 65, 27, 20, 36, 0, 30, 42, 71, 52...   
Fault                                      [Policy Holder, Third Party]   
PolicyType            [Sport - Liability, Sport - Collision, Sedan -...   
VehicleCategory                                 [Sport, Utility, Sedan]   
VehiclePrice          [more than 69000, 20000 to 29000, 30000 to 390...   
FraudFound_P                                                     [0, 1]   
PolicyNumber          [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   
RepNumber             [12, 15, 7, 4, 3, 14, 1, 13, 11, 16, 6, 2, 8, ...   
Deductible                                         [300, 400, 500, 700]   
DriverRating                                               [1, 4, 3, 2]   
Days_Policy_Accident    [more than 30, 15 to 30, none, 1 to 7, 8 to 15]   
Days_Policy_Claim               [more than 30, 15 to 30, 8 to 15, none]   
PastNumberOfClaims                       [none, 1, 2 to 4, more than 4]   
AgeOfVehicle          [3 years, 6 years, 7 years, more than 7, 5 yea...   
AgeOfPolicyHolder     [26 to 30, 31 to 35, 41 to 50, 51 to 65, 21 to...   
PoliceReportFiled                                             [No, Yes]   
WitnessPresent                                                [No, Yes]   
AgentType                                          [External, Internal]   
NumberOfSuppliments                 [none, more than 5, 3 to 5, 1 to 2]   
AddressChange_Claim   [1 year, no change, 4 to 8 years, 2 to 3 years...   
NumberOfCars          [3 to 4, 1 vehicle, 2 vehicles, 5 to 8, more t...   
Year                                                 [1994, 1995, 1996]   
BasePolicy                           [Liability, 

**Observations👁️:**

- **Data Type 🧩:** 
The dataset includes a mix of numeric and categorical features, with some Boolean variables that need conversion for better
modeling.

- **N Unique 🔢:** Each feature has a varying number of unique values. Some features, like "Sex" or "Fault," have only two unique categories, while 

   others may have more complex value distributions.

- **Unique Values 🆚**: Features like "PolicyNumber" or "AgentType" may have repetitive values, while others, like "Age," present more diversity.

- **Max & Min Values ⬆️⬇️**: Some features, like "Age," have reasonable ranges, but others, such as "Age" with a minimum value of 0, require 

  attention to ensure data validity.

- **Measures of Central Tendency 📏:** Features like "Age" and "DriverRating" show typical values (mean, median, mode) that can inform the 

    distribution of the data.

- **Range, Variance, Standard Deviation 📐:** Numeric features like "VehiclePrice" have a wide range and considerable variance, suggesting that they

   could be crucial for predicting fraud.

- **Kurtosis 📊:** Features with high kurtosis, like "DriverRating," indicate heavy tails in their distributions, which might suggest the presence of 

    extreme values.

- **Skewness 📉:**  Certain features, such as "FraudFound_P" and "Deductible," exhibit skewed distributions, which might affect model accuracy and 

    need transformation.

#### **2.0 ✨ Exploratory Data Analysis (EDA) 🔍**

**Why is EDA Necessary?**


**Data Understanding:**  🧠  EDA provides insights into the dataset, helping to understand its structure, distribution, and inherent patterns.

**Problem Indentification:**  🚩 It helps identify issues such as missing values, outliers, and anomalies that need to be addressed during data preprocessing.

**Informing Modeling Strategies:** 🛠️📈   Understanding the data helps in choosing the appropriate modeling techniques and preprocessing steps.

**Insight Discovery:** 💡🔗 EDA can reveal hidden patterns and relationships within the data, providing valuable insights for decision-making and 

strategy formulation

In [10]:
df_viz = data.copy()
df_viz.shape

(15420, 33)

#### **2.1 🎯 Target Balance Visualization** 📊


Target👉**FraudFound_P**: Indicates whether fraud was detected in the insurance claim.

In [11]:
Fraud_count= df_viz['FraudFound_P'].value_counts().reset_index()
Fraud_count.columns =['FraudFound_P','Count']
Fraud_count


FraudFound_P  Count
0             0  14497
1             1    923

**Visualizing the target variable (binary: 0 and 1) as a pie chart 🥧**


In [12]:
fig = px.pie(Fraud_count, names='FraudFound_P', values='Count', color='FraudFound_P',color_discrete_map={0: '#B0E0E6', 1: '#FFB6C1'})
fig.update_traces(
    textinfo='percent',
    textfont={'size':16,'color':'Black'},
    marker=dict(line=dict(color='black', width=2)))
fig.update_layout(
    title={
        'text': 'Target Balance',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.show()

**Observations on Target Variable 🚦**

- **Fraud_Found_P is Highly Imbalanced** 📊  


























  - Majority class dominates the dataset, causing imbalance.

**Issues with Imbalanced Target Variable ⚠️**  

1. **Model Bias**: Skewed learning towards the majority class 🧠➡️⚪.  

2. **Misleading Metrics**: High accuracy may just reflect majority predictions 📈❌.  

3. **Poor Generalization**: Struggles with minority class in real-world scenarios 🌍🤔.


**Why Oversampling? 🌟**

- **Improves Model Performance**: Helps predict minority class better 🏆📉. 

- **Enhances Evaluation Metrics**: Boosts Precision, Recall, and F1-Score accuracy 🔍✔️.

- **Balances Learning**: Ensures fair learning from both classes ⚖️📚.  


👉 **Next Step**: Perform **Oversampling** during **Data Preprocessing** 🔄✨.


#### **2.2 🔍 Analyzing Relationships Between Target and Features 📊**

##### **2.2.1  🚨 Fraud Detection by Sex 🚻**

In [13]:
df_fraud=df_viz[df_viz['FraudFound_P']==1]
fraud_counts_by_sex=df_fraud['Sex'].value_counts()
fraud_counts_by_sex

Sex
Male      818
Female    105
Name: count, dtype: int64

In [14]:
fraud_percentages_by_sex=(fraud_counts_by_sex/fraud_counts_by_sex.sum())*100
fraud_percentages_by_sex

Sex
Male      88.624052
Female    11.375948
Name: count, dtype: float64

In [15]:
#Bar chart 
fig = go.Figure()

fig.add_traces(go.Bar(x=fraud_counts_by_sex.index, y=fraud_counts_by_sex.values,
                      text=[f'{count} ({percentage:.2f}%)' for count, percentage in zip(fraud_counts_by_sex.values,fraud_percentages_by_sex)],
                      textposition='auto',marker_color=['rgba(31, 119, 180, 0.6)', 'rgba(255, 127, 14, 0.6)'], width=0.5))
fig.update_layout(
    title={
    'text':'Fraud Detection by Sex',
    'x': 0.5,
    'font':{'family':"Arial, sans-serif",'size':24}
    })

fig.update_layout(
    xaxis=dict(title='Sex'),
    yaxis=dict(title='Count'),
    showlegend=False)

fig.show()

This bar chart displays the percentage distribution of males and females in the dataset, with **males representing 88.62%** and **females representing 11.38%.**

👉The results suggest that **males** are significantly more likely to be involved in detected fraud cases compared to **females**. 🚹📊🚺


##### **2.2.2 Fraud Detection by Age** 🧑‍🤝‍🧑📊  **& Age of Policy Holder** 🧓👵🔍


In [16]:
df_fraud = df_viz[df_viz['FraudFound_P'] == 1]
df_non_fraud = df_viz[df_viz['FraudFound_P'] == 0]

df_counts_age = df_viz['Age'].value_counts().sort_index()
df_counts_fraud = df_fraud['Age'].value_counts().sort_index()

df_percentages_fraud = round((df_counts_fraud / df_counts_age) * 100, 2)

# Bar Chart
fig = go.Figure()

fig.add_trace(go.Bar(x=df_percentages_fraud.index, y=df_percentages_fraud.values,
    marker_color='rgba(31, 119, 180, 0.7)'))

fig.update_layout(
    title={
        'text': 'Fraud Detection by Age',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.update_layout(
    xaxis=dict(title='Age'), yaxis=dict(title='Percentage', range=[0, 20], dtick=2), legend=dict(title='Outcome'))

fig.add_shape(
    type="line",
    x0=min(df_percentages_fraud.index),
    y0=6,
    x1=max(df_percentages_fraud.index),
    y1=6,
    line=dict(
        color="rgba(255, 0, 0, 0.8)",
        width=2,
        dash="dashdot",
    ),
)

fig.show()

In [17]:
df_fraud = df_viz[df_viz['FraudFound_P'] == 1]
df_non_fraud = df_viz[df_viz['FraudFound_P'] == 0]

df_counts_ageofpolicyholder = df_viz['AgeOfPolicyHolder'].value_counts().sort_index()
df_counts_fraud_ = df_fraud['AgeOfPolicyHolder'].value_counts().sort_index()

df_percentages_fraud_ = round((df_counts_fraud_ / df_counts_ageofpolicyholder) * 100, 2)

# Bar Chart
fig = go.Figure()

fig.add_trace(go.Bar(x=df_percentages_fraud_.index, y=df_percentages_fraud_.values,
    marker_color='rgba(31, 119, 180, 0.7)'))

fig.update_layout(
    title={
        'text': 'Fraud Detection by Age of Policy Holder',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.update_layout(
    xaxis=dict(title='Age of Policy Holder'), yaxis=dict(title='Percentage', range=[0, 20], dtick=2), legend=dict(title='Outcome'))

fig.add_shape(
    type="line",
    x0=-1,
    y0=6,
    x1=10,
    y1=6,
    line=dict(
        color="rgba(255, 0, 0, 0.8)",
        width=2,
        dash="dashdot",
    ),
)

fig.show()


**Age:** The age of the policyholder.

**AgeOfPolicyHolder:** The age of the policyholder.

##### **2.2.3 Fraud Detection by Make** 🚗🔍📊

In [18]:
df_counts_make = df_viz['Make'].value_counts().sort_index()
df_counts_fraud2 = df_fraud['Make'].value_counts().sort_index()
df_percentages_fraud2 = pd.DataFrame(round((df_counts_fraud2 / df_counts_make) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud2.columns = ['Make', 'Fraud %']
df_percentages_fraud2 = df_percentages_fraud2.sort_values(by=['Fraud %', 'Make'])

# Bar Chart
fig = go.Figure()

fig.add_trace(go.Bar(x=df_percentages_fraud2['Fraud %'], y=df_percentages_fraud2['Make'], orientation='h',
    marker_color='lightcoral'))

fig.update_layout(
    title={
        'text': 'Fraud Detection by Make',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.update_layout(
    xaxis=dict(title='Percentage', range=[0, 30], dtick=2), yaxis=dict(title='Make'), legend=dict(title='Outcome'), bargap=0.3)

fig.add_shape(
    type="line",
    x0=6,
    y0=0,
    x1=6,
    y1=len(df_percentages_fraud2['Make']) - 1,
    line=dict(
        color="rgba(128, 128, 128, 1)",
        width=2,
        dash="dashdot", 
    ),
)

fig.show()

##### **2.2.4 Fraud Detection by Vehicle Price** 🚗💰🔍📊  

In [19]:
df_counts_vp = df_viz['VehiclePrice'].value_counts().sort_index()
df_counts_fraud3 = df_fraud['VehiclePrice'].value_counts().sort_index()
df_percentages_fraud3 = pd.DataFrame(round((df_counts_fraud3 / df_counts_vp) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud3.columns = ['VehiclePrice', 'Fraud %']
df_percentages_fraud3 = df_percentages_fraud3.sort_values(by=['Fraud %'])

# Bar Chart
fig = go.Figure()

fig.add_trace(go.Bar(x=df_percentages_fraud3['Fraud %'], y=df_percentages_fraud3['VehiclePrice'], orientation='h',
    marker_color='#9FF781'))

fig.update_layout(
    title={
        'text': 'Fraud Detection by VehiclePrice',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.update_layout(
    xaxis=dict(title='Percentage', range=[0, 16], dtick=2), yaxis=dict(title='VehiclePrice'), legend=dict(title='Outcome'), bargap=0.3)

fig.add_shape(
    type="line",
    x0=6,
    y0=0,
    x1=6,
    y1=5.5,
    line=dict(
        color="rgba(0, 100, 0, 1)",
        width=2,
        dash="dashdot", 
    ),
)

fig.show()

##### **2.2.5 Fraud Detection by Base Policy** 🛡️📑🔍 **& Policy Type** 📝🚗📊  

In [20]:
# BasePolicy
df_counts_policy = df_viz['BasePolicy'].value_counts().sort_index()
df_counts_fraud4 = df_fraud['BasePolicy'].value_counts().sort_index()
df_percentages_fraud4 = pd.DataFrame(round((df_counts_fraud4 / df_counts_policy) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud4.columns = ['BasePolicy', 'Fraud %']
df_percentages_fraud4 = df_percentages_fraud4.sort_values(by=['Fraud %'])

# PolicyType
df_counts_policy2 = df_viz['PolicyType'].value_counts().sort_index()
df_counts_fraud5 = df_fraud['PolicyType'].value_counts().sort_index()
df_percentages_fraud5 = pd.DataFrame(round((df_counts_fraud5 / df_counts_policy2) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud5.columns = ['PolicyType', 'Fraud %']
df_percentages_fraud5 = df_percentages_fraud5.sort_values(by=['Fraud %'])

# Bar Chart - Subplots
fig = make_subplots(rows=1, cols=2, subplot_titles=('BasePolicy', "PolicyType"))

fig.add_trace(go.Bar(x=df_percentages_fraud4['BasePolicy'], y=df_percentages_fraud4['Fraud %'],
    marker_color='skyblue'), row=1, col=1)

fig.add_trace(go.Bar(x=df_percentages_fraud5['PolicyType'], y=df_percentages_fraud5['Fraud %'],
    marker_color='coral'), row=1, col=2)

fig.update_layout(
    title={
        'text': 'Fraud Detection by BasePolicy & PolicyType',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    }, showlegend=False)

fig.add_shape(
    type="line",
    x0=-1,
    y0=6,
    x1=3,
    y1=6,
    line=dict(
        color="rgba(128, 128, 128, 1)", 
        width=2,
        dash="dashdot",
    ),
    xref="x1",
    yref="y1"
)

fig.add_shape(
    type="line",
    x0=-1,
    y0=6,
    x1=9,
    y1=6,
    line=dict(
        color="rgba(128, 128, 128, 1)", 
        width=2,
        dash="dashdot",
    ),
    xref="x2",
    yref="y2"
)

fig.show()


##### **2.2.6 Fraud Detection by Age of Vehicle** 🚗⏳🔍📊  


In [21]:
# Age of Vehicle
df_counts_ageofvehicle = df_viz['AgeOfVehicle'].value_counts().sort_index()
df_counts_fraud6 = df_fraud['AgeOfVehicle'].value_counts().sort_index()
df_percentages_fraud6 = pd.DataFrame(round((df_counts_fraud6 / df_counts_ageofvehicle) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud6.columns = ['AgeOfVehicle', 'Fraud %']
df_percentages_fraud6 = df_percentages_fraud6.sort_values(by=['Fraud %'])

# Bar Chart
fig = go.Figure()

fig.add_trace(go.Bar(x=df_percentages_fraud6['Fraud %'], y=df_percentages_fraud6['AgeOfVehicle'], orientation='h',
    marker_color='rgb(255, 165, 0)'))

fig.update_layout(
    title={
        'text': 'Fraud Detection by Age of Vehicle',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    })

fig.update_layout(
    xaxis=dict(title='Percentage', range=[0, 16], dtick=2), yaxis=dict(title='AgeOfVehicle'), legend=dict(title='Outcome'), bargap=0.3)

fig.add_shape(
    type="line",
    x0=6,
    y0=-1,
    x1=6,
    y1=8,
    line=dict(
        color="rgba(0, 100, 0, 1)",
        width=2,
        dash="dashdot", 
    ),
)

fig.show()

##### **2.2.7 Fraud Detection by Month** 📅🔍 **& Month Claimed** 🗓️🚗📊  


In [22]:
# Month
df_counts_month = df_viz['Month'].value_counts().sort_index()
df_counts_fraud7 = df_fraud['Month'].value_counts().sort_index()
df_percentages_fraud7 = pd.DataFrame(round((df_counts_fraud7 / df_counts_month) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud7.columns = ['Month', 'Fraud %']

# MonthClaimed
df_counts_monthclaimed = df_viz['MonthClaimed'].value_counts().sort_index()
df_counts_fraud8 = df_fraud['MonthClaimed'].value_counts().sort_index()
df_percentages_fraud8 = pd.DataFrame(round((df_counts_fraud8 / df_counts_monthclaimed) * 100, 2)).fillna(0).reset_index()

df_percentages_fraud8.columns = ['MonthClaimed', 'Fraud %']

month_order = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

# Bar Chart - Subplots
fig = make_subplots(rows=1, cols=2, subplot_titles=('Month', "MonthClaimed"))

fig.add_trace(go.Bar(x=df_percentages_fraud7['Month'], y=df_percentages_fraud7['Fraud %'],
    marker_color='#8A2BE2'), row=1, col=1)

fig.add_trace(go.Bar(x=df_percentages_fraud8['MonthClaimed'], y=df_percentages_fraud8['Fraud %'],
    marker_color='#FF4500'), row=1, col=2)

fig.update_layout(
    title={
        'text': 'Fraud Detection by Month & MonthClaimed',
        'x':0.5,
        'font': {'family': "Arial, sans-serif", 'size': 24}
    }, showlegend=False,
    xaxis=dict(categoryorder='array', categoryarray=month_order),
    xaxis2=dict(categoryorder='array', categoryarray=month_order))

fig.add_shape(
    type="line",
    x0=-1,
    y0=6,
    x1=12,
    y1=6,
    line=dict(
        color="rgba(128, 128, 128, 1)", 
        width=2,
        dash="dashdot",
    ),
    xref="x1",
    yref="y1"
)

fig.add_shape(
    type="line",
    x0=-1,
    y0=6,
    x1=12.5,
    y1=6,
    line=dict(
        color="rgba(128, 128, 128, 1)", 
        width=2,
        dash="dashdot",
    ),
    xref="x2",
    yref="y2"
)

fig.show()

**MonthClaimed:** The month in which the insurance claim was processed.

**Month:** The month in which the insurance claim was made.

**Month** and **MonthClaimed**  have similar distributions =>We don't need both